### DISCUSSION TITLE
Modifying ONNX graph

### DISCUSSION THREAD

**Discussion by ChrisPB3** [onnx/onnx#4559] [Q&A] [answered]
I am trying to create a new input and subsequent node in the ONNX graph using the onnx.helper utility. I am able to correctly create the input tensor and node and insert the node into the graph.node list object. If I visualize the model in Netron it appears exactly as I want it but when using the onnx.checker.checkmodel() function it produces the issue of 'Nodes in a graph must be topologically sorted, however input 'tensor_name' of node: name: 'node_name' OpType: Conv is not output of any previous nodes. This issue does not appear if I insert new nodes in between the model input and the original 'input' node.
Please help me with what aspect I need to include in order to maintain the model's topological order when creating these extra inputs and subsequent node structures.

**Comment by jcwchen** [comment]
Hi @ChrisPB3,
ONNX IR requires ONNX nodes to be topological-sorted in `model.graph.node`. That is to say, the inputs for a node can only come from previous nodes instead of later nodes.

A possible solution would be inserting all of the nodes in the first place regardless any order and then use their direction relationship (a -> b) to reorder existing nodes with topological sort.

**Reply by ChrisPB3** [reply, marked as answer]
Thank you for the answer. I actually managed to get to get it working. There was an issue with how the input tensor was being created but once I fixed that the topological order issue was also fixed. The node taking the input tensor was looking for a node in the network corresponding to the input tensor name and because there was no such node it produced this error.

**Comment by zhenhuaw-me** [comment]
@ChrisPB3 You can try if https://github.com/ZhangGe6/onnx-modifier works for you. I believe it should do the topoligical sorting for you.

**Reply by ChrisPB3** [reply]
Thank you for showing me this tool! However, I am trying to set up something more powerful with more possibilities of editing the ONNX model. Luckily I managed to fix the original issue but thanks for the help!

**Comment by sirine-sirine** [comment]
hello i have this error when i want to use my model.onnx to do some predictions 
TypeError: run(): incompatible function arguments. The following argument types are supported:
    1. (self: onnxruntime.capi.onnxruntime_pybind11_state.InferenceSession, arg0: List[str], arg1: Dict[str, object], arg2: onnxruntime.capi.onnxruntime_pybind11_state.RunOptions) -> List[object]

Can you please help me 
thank you

**Comment by Manueljohnson063** [comment]
Hai @jcwchen 
I am currently working with a large ONNX model (main.onnx) and need to replace a specific section (referred to as a subgraph) within it. I intend to use a smaller ONNX model (sub.onnx) as the replacement for this section. Essentially, I want to integrate sub.onnx into the appropriate section of main.onnx.

The below image main.onnx is a portion from a large graph ,I want to place the sub.onnx in to the main.onnx(ie the name of input to the reshape is **name: Add.181**  after matmul is **name: matmul_v2_1.tmp_0**   ,
TIA

The main.onnx :+1: 
![image](https://github.com/user-attachments/assets/5db60b91-c4e5-41a9-ab72-d410a7b7cb0b)



The sub.onnx model :
![subgraph](https://github.com/user-attachments/assets/60d33547-f2fc-4ecb-a118-63bbf856ceb9)
